### ISSUE TITLE
`ClipboardSetText` successfully called, but actually failed

### ISSUE THREAD

**Issue by Licoy** [wailsapp/wails#2534]
### Description

On Macbook M1 Pro, after building, the ClipboardSetText is called. The feedback is successful, but the actual clipboard is empty

### To Reproduce
```js
ClipboardSetText(str).then(()=>{
    console.log('success')
  }).catch((err)=>{
    console.err('fail',err)
})

```
### Expected behaviour

It should be possible to paste directly

### System Details

```shell
# System

OS           | MacOS    
Version      | 13.2.1     
Go Version   | go1.18.10
Platform     | darwin   
Architecture | arm64    

# Wails

Version | v2.4.1

# Dependencies

Dependency                | Package Name | Status    | Version     
Xcode command line tools  | N/A          | Installed | 2396        
npm                       | N/A          | Installed | 8.1.2       
*Xcode                    | N/A          | Installed | 14.2 (14C18)
*upx                      | N/A          | Available |             
*nsis                     | N/A          | Available |             
* - Optional Dependency
```

**Comment by leaanthony** [comment]
If you don't have paste in your menu with a cmd-v accelerator, you won't be able to paste

**Comment by CiroLee** [comment]
I encountered the same problem, using the api provided by wails, the dev mode is able to copy and paste, after the build, sometimes you can paste, sometimes not, it is puzzling

**Comment by zveinn** [comment]
Can confirm this issue still exists, copy and paste commands are visible in the toolbar menu. Paste always returns and empty text and copy does not seem to work at all. Old clipboard values are still present after copy.

**Comment by ethanwillis** [comment]
The main culprit for *no* data being returned is probably the pasteboard* daemon dying (see: [[3](https://apple.stackexchange.com/a/111076)]). If that daemon dies and set and gets/copies + pasts won't work properly. And even if the daemon comes back up the app will maybe need to be completely restarted. (*pasteboard is what Apple calls a clipboard)

Inconsistencies could also be due to how `pbpaste` works when there are multiple items in various formats stored in the clipboard. All the wails implementations are using the system's general pasteboard either implicitly or explicitly in wails v2 or wails v3 respectively. 

**In wails v2...**(see: [[5](https://github.com/wailsapp/wails/blob/2859143b2f9966dd3fb2096744feed95dc78c195/v2/internal/frontend/desktop/darwin/clipboard.go#L9)])  
  The `pbcopy` command is called directly via exec and piped input over stdin putting whatever text is given into the general pasteboard.  

  The `pbpaste` command is called the same way and the output from stdout is returned. However, no additional options for `pbpaste` are used. If you look at the manpage for `pbpaste` you'll notice `-Prefer` as an option. 

  This option is not used explicitly in the wails v2 implementation. However, the reason this option exists is because a pasteboard can have multiple **types** of data stored in it (see: [[7](https://developer.apple.com/library/archive/documentation/General/Devpedia-CocoaApp-MOSX/Pasteboard.html#//apple_ref/doc/uid/TP40009448-CH21-SW1), [8](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/PasteboardGuide106/Articles/pbGettingStarted.html#//apple_ref/doc/uid/TP40008150-SW1)]). `pbpaste` will check the pasteboard for a representation of each of these types of data. If none are present it will return **nothing**. 

And in general there are **other known problems** (see: [[4](https://stackoverflow.com/a/68971320)]) with getting back rich text ***as*** rich text.

TL;DR It's possible that...
- You can get back data from the clipboard with "data loss". (Ex. Rich text becoming ASCII plaintext).
- You can get back **no** data from clipboard when the **type** of data stored in he clipboard is not ASCII, PS, or RTF data. 
- You can get back old/incorrect/no data from the clipboard because the pboard(clipboard/pasteboard) daemon has died/restarted. 

**In wails v3... I refer you to the implementation in [[6](https://github.com/wailsapp/wails/blob/2859143b2f9966dd3fb2096744feed95dc78c195/v3/pkg/application/clipboard_darwin.go#L35)] using the information from above**

**Refs:**
1. [man pbpaste](https://ss64.com/osx/pbpaste.html)
2. [man pbcopy](https://ss64.com/osx/pbcopy.html)
3. [restarting pboard daemon + apps](https://apple.stackexchange.com/a/111076)
4. [rich text coming out of pasteboard as plain text](https://stackoverflow.com/a/68971320)
5. [wails v2 darwin clipboard implementation](https://github.com/wailsapp/wails/blob/2859143b2f9966dd3fb2096744feed95dc78c195/v2/internal/frontend/desktop/darwin/clipboard.go#L9)
6. [wails v3 darwin clipboard implementation](https://github.com/wailsapp/wails/blob/2859143b2f9966dd3fb2096744feed95dc78c195/v3/pkg/application/clipboard_darwin.go#L35)
7. [Apple Pasteboard overview](https://developer.apple.com/library/archive/documentation/General/Devpedia-CocoaApp-MOSX/Pasteboard.html#//apple_ref/doc/uid/TP40009448-CH21-SW1)
8. [Apple Pasteboard programmer guide](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/PasteboardGuide106/Articles/pbGettingStarted.html#//apple_ref/doc/uid/TP40008150-SW1)

**Comment by leaanthony** [comment]
Yeah the v3 API was changed from get/set to explicitly gettext/settext for this reason. Im assuming nothing can be done for Linux as there's no standard API for it?

**Comment by leaanthony** [comment]
> It still seems kind of strange that the clipboard would work just fine when running "wails dev" but then breaks on the same text once the app is built. Maybe there are some implementation details that differ between the actual build and the dev environment ?

Odd. The only platform that may exhibit this behaviour is macOS as the plist can often affect functionality

**Comment by zveinn** [comment]
I tried using the same .plist file for production and dev but no dice. 

It seems like there is a long running bug on webkit: https://bugs.webkit.org/show_bug.cgi?id=75891
Here is a stackoverflow thread talking about some possible fixes: https://stackoverflow.com/questions/29376510/custom-cut-copy-paste-operations-for-wkwebview
This thread references the copy/paste javascript events inside webkit as a possible workaround: https://developer.apple.com/library/archive/documentation/AppleApplications/Conceptual/SafariJSProgTopics/CopyAndPaste.html

**Comment by zveinn** [comment]
I just noticed something else.. the entire "right click" context menu is disabled after building but works during dev runs. 

Maybe there is some code that runs during production build that is trying to intercept/overwrite/modify the context menu during production builds that's causing the context menu to break, which would then cause copy/paste to not work, since it's most likely tied to the context menu

@leaanthony

**Comment by leaanthony** [comment]
Master now has the fixes for this. Please test using [this guide](https://wails.io/docs/guides/local-development). If it doesn't work as expected, please reopen 👍

**Comment by zveinn** [comment]
I will give it a go once my coffee kicks in, thanks !

**Comment by zveinn** [comment]
Hey, I tested out the main branch and it does indeed work. 

BUT.. there is a potential bug/unintended functionality regarding osascript that was breaking the clipboard for me. 

I was using osascript with a prompt, kind of like this 
```bash
osascript -e 'do shell script "./NicelandVPN" with prompt "Run NicelandVPN as admin" with administrator privileges'
```
The problem that was occurring is that the GUI prompt which let's you insert your username/password is grabbing the stdin/TTY (which is used by the clipboard to deliver text) and it was not returning it to the binary being executed by the osascript. 

This lead me to abandon the osascript path and instead writing a C file that uses the macos internal Security Framework to generate admin credentials. 

I will make the C code available via a git repository tomorrow if you want to use it in order to enable users to launch wails.io apps as administrator with a GUI prompt.

**Comment by leaanthony** [comment]
This would make an ideal v3 plugin

**Comment by zveinn** [comment]
@leaanthony here you go: https://github.com/zveinn/macos-app-builder
There are scripts in their for codesigning, notarization, icon generation. 

the launcher.c is the automatice launcher code.

**Comment by timmattison** [comment]
Can you repost the video? It is unavailable on YouTube now.

**Comment by leaanthony** [comment]
Looks like this can be closed now.
